File: python/y2023/d08/aoc_2023_08.py
import functools
import math
from itertools import cycle

from common import Puzzle, timer


class Day08(Puzzle, year=2023, day=8):
    @staticmethod
    def clean_input(data):
        network = dict()
        for node in data[2:]:
            network[node[0:3]] = (node[7:10], node[12:15])
        return_data = (data[0], network)
        return return_data

    @timer(part=1)
    def solve_part_1(self):
        """Solution for part 1"""
        steps = cycle(self.data[0])
        step_index = 0
        found = False
        map = self.data[1]

        destination = map["AAA"]
        while not found:
            step = next(steps)
            step_index += 1
            if step == "R":
                new_node = destination[1]
            else:
                new_node = destination[0]
            if new_node == "ZZZ":
                found = True
            else:
                destination = map[new_node]

        return step_index

    @staticmethod
    def get_steps(node, instructions, network):
        steps_required = 0
        steps = cycle(instructions)
        step_index = 0
        while True:
            step = next(steps)
            step_index += 1
            if step == "R":
                node = network[node][1]
            else:
                node = network[node][0]

            if node.endswith("Z"):
                steps_required = step_index
                break

        return steps_required

    @timer(part=2)
    def solve_part_2(self):
        """Solution for part 2"""
        steps_required = dict()
        instructions = self.data[0]
        network = self.data[1]
        nodes = [node for node in network if node.endswith("A")]
        for node in nodes:
            steps_required[node] = Day08.get_steps(node, instructions, network)

        return functools.reduce(math.lcm, steps_required.values())

File: C++/2023/08/aoc_2023_08.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <algorithm>
#include <map>
#include <numeric>
#include <string>
#include <vector>

namespace aoc_2023_08 {

using node_t = std::pair<std::string, std::string>;
using network_t = std::map<std::string, node_t>;
using instructions_t = std::string;
using aoc_data_t = std::vector<std::string>;

std::pair<instructions_t, network_t> clean_input(aoc_data_t input_data) {
    network_t      network;
    instructions_t instructions = input_data[0];

    for (aoc_data_t::iterator it = input_data.begin() + 2; it < input_data.end(); it++) {
        auto        row = *it;
        std::string node = row.substr(0, 3);

        std::pair<std::string, std::string> destination;

        destination.first = row.substr(7, 3);
        destination.second = row.substr(12, 3);
        network[node] = destination;
    }
    return std::make_pair(instructions, network);
}

std::int64_t get_steps(std::string node, instructions_t instructions, network_t network, int part) {
    std::int64_t steps_required = 0;

    char step;

    while (true) {
        char step = instructions[0];
        std::rotate(instructions.begin(), instructions.begin() + 1, instructions.end());
        steps_required++;
        if (step == 'R') {
            node = network[node].second;
        } else {
            node = network[node].first;
        }
        if (part == 1) {
            if (node == "ZZZ") {
                break;
            }
        } else {
            if (node[2] == 'Z') {
                break;
            }
        }
    }
    return steps_required;
}

std::int64_t solve_part_1(const std::pair<instructions_t, network_t> &data) {
    std::int64_t steps = 0;

    std::string    node = "AAA";
    instructions_t instructions = data.first;
    network_t      network = data.second;

    return get_steps(node, instructions, network, 1);
}

std::int64_t lcm(std::int64_t a, std::int64_t b) {
    return (a * b) / (std::gcd(a, b));
}

std::int64_t solve_part_2(const std::pair<instructions_t, network_t> &data) {
    std::int32_t   prio_sum = 0;
    instructions_t instructions = data.first;
    network_t      network = data.second;

    std::vector<std::int64_t> steps_required;

    for (auto &node : network) {
        if (node.first.ends_with("A")) {
            steps_required.push_back(get_steps(node.first, instructions, network, 2));
        }
    }

    return std::accumulate(steps_required.begin(), steps_required.end(), steps_required[0], std::lcm<std::int64_t, std::int64_t>);
}

void *solve_all(const std::pair<instructions_t, network_t> &data) {
    aoc::timer(1, solve_part_1, data, 1);
    aoc::timer(2, solve_part_2, data, 1);
    return NULL;
}

} // namespace aoc_2023_08

int main(int argc, char **argv) {
    std::string filename;
    const int   year = 2023;
    const int   day = 8;

    if (argc > 1) {
        if (std::string(argv[1]) == "--test") {
            filename = "test_input.txt";
        } else {
            filename = argv[1];
        }
    } else {
        filename = "input.txt";
    }

    std::vector<std::string> data = aoc::io::get_input_list<std::string>(filename, year, day);

    auto parsed_data = aoc_2023_08::clean_input(data);
    printf("================================================\n");
    printf("Solution for %d, day %02d\n", year, day);
    aoc::timer(0, aoc_2023_08::solve_all, parsed_data, 0);

    return 0;
}
